File: tp_tests/logreg.py
def logreg(data=listf100, g=listf100, w=listf100, dims=int): 
    dot = 1.0
    j = 0
    while(j<=dims):
        dot += data[j]*w[j]
        j+=1
    label = data[dims]
    scale = (1.0 / (1.0+ exp(-label*dot)) - 1.0) * label
    j2 = 0
    while(j2<=dims):
        g[j2]+=scale*data[j2]
        j2+=1

def main():
    data = range(100.0)
    data[99]=-1.0
    g = zeros(100.0)
    w = range(100.0)
    print g
    logreg(data, g, w, 99)
    print g

File: tp_tests/llvm_clang/logreg.cpp

#import <stdio.h>
#import <math.h>

void logreg(float* data, float* g, float* w, int DIMS) {
    float dot = 0;
    for (int j = 0; j < DIMS; j++)
        dot += data[j] * w[j];
    float label = data[DIMS];
    float scale = (1 / (1 + exp(-label * dot)) - 1) * label;
    for (int j = 0; j < DIMS; j++)
        g[j] += scale * data[j];
}

int main() {
    float* data = new float[100];
    float* w = new float[100];
    float* g = new float[100];

    for(int i=0;i<100;i++) {
        data[i]=i;
        w[i]=i;
        g[i]=0.0;
    }
    data[99]=-1.0;
    logreg(data, g, w, 99);
    /* 
    printf("[");
    for(int i=0;i<100;i++) {
        printf("%f, ", g[i]);
    }
    printf("]\n");
    */
    delete[] data;
    delete[] w;
    delete[] g;
}
